pkg/volume/iscsi/iscsi_util.go

```diff
@@ -263,66 +254,92 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
 			if err != nil {
 				return err
 			}
-			// Extract the iface from the mountPath and use it to log out. If the iface
-			// is not found, maintain the previous behavior to facilitate kubelet upgrade.
-			// Logout may fail as no session may exist for the portal/IQN on the specified interface.
-			iface, found := extractIface(mntPath)
-			if found {
-				glog.Infof("iscsi: log out target %s iqn %s iface %s", portal, iqn, iface)
-				out, err := c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "-I", iface, "--logout"})
-				if err != nil {
-					glog.Errorf("iscsi: failed to detach disk Error: %s", string(out))
-				}
-				// Delete the node record
-				glog.Infof("iscsi: delete node record target %s iqn %s", portal, iqn)
-				out, err = c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "-I", iface, "-o", "delete"})
+			refCount, err := getDevicePrefixRefCount(c.mounter, prefix)
+			if err == nil && refCount == 0 {
+				// This portal/iqn/iface is no longer referenced, log out.
+				// Extract the portal and iqn from device path.
+				portal, iqn, err := extractPortalAndIqn(device)
 				if err != nil {
-					glog.Errorf("iscsi: failed to delete node record Error: %s", string(out))
+					return err
 				}
-			} else {
-				glog.Infof("iscsi: log out target %s iqn %s", portal, iqn)
-				out, err := c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "--logout"})
-				if err != nil {
-					glog.Errorf("iscsi: failed to detach disk Error: %s", string(out))
-				}
-				// Delete the node record
-				glog.Infof("iscsi: delete node record target %s iqn %s", portal, iqn)
-				out, err = c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "-o", "delete"})
-				if err != nil {
-					glog.Errorf("iscsi: failed to delete node record Error: %s", string(out))
+				// Extract the iface from the mountPath and use it to log out. If the iface
+				// is not found, maintain the previous behavior to facilitate kubelet upgrade.
+				// Logout may fail as no session may exist for the portal/IQN on the specified interface.
+				iface, found := extractIface(mntPath)
+				if found {
+					glog.Infof("iscsi: log out target %s iqn %s iface %s", portal, iqn, iface)
+					out, err := c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "-I", iface, "--logout"})
+					if err != nil {
+						glog.Errorf("iscsi: failed to detach disk Error: %s", string(out))
+					}
+					// Delete the node record
+					glog.Infof("iscsi: delete node record target %s iqn %s", portal, iqn)
+					out, err = c.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", portal, "-T", iqn, "-I", iface, "-o", "delete"})
+					if err != nil {
+						glog.Errorf("iscsi: failed to delete node record Error: %s", string(out))
+					}
+				} else {
+					glog.Infof("iscsi: log out target %s iqn %s", portal, iqn)
```

Comment:
@rootfs how long backwards compatibility is promised.
Can this code be removed after X versions of k8s.
If so would be good to add a todo for when it should be removed